### PR TITLE
[16.0][MIG] migrate purchase_cancel_reason

### PR DESCRIPTION
**Pull request by pierre-halleux**

### CODE REVIEW

File: purchase_cancel_reason/model/purchase_order.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo import fields, models
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    cancel_reason_id = fields.Many2one(
+        comodel_name="purchase.order.cancel.reason",
+        string="Reason for cancellation",
```
Comment (lmignon) [inline purchase_cancel_reason/model/purchase_order.py]:
```suggestion
        string="Cancellation reason",
```

File: purchase_cancel_reason/views/purchase_order.xml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                />
+            </xpath>
+            <xpath expr="/form/sheet/div/h1" position="after">
+                <h5 class="mb-3" attrs="{'invisible': [('state', '!=', 'cancel')]}">
+                    <label for="cancel_reason_id" string="Cancellation reason:" />
+                    <field
+                        name="cancel_reason_id"
+                        class="oe_inline"
+                        options='{"no_open": True}'
+                    />
```
Comment (lmignon) [inline purchase_cancel_reason/views/purchase_order.xml]: I propose to improve the view as follow:
```suggestion
                     <label for="cancel_reason_id"/>
                     <pan> </pan>
                    <field name="cancel_reason_id" class="oe_inline" options="{&quot;no_open&quot;: True}"/>
```
Before:
image

After:
image


After

**Review by lmignon — CHANGES_REQUESTED** [review]
@pierre-halleux Some little details...

**Review by lmignon — APPROVED** [review]
LGTM (Code review + functional test)

File: purchase_cancel_reason/model/purchase_order.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        readonly=True,
+        ondelete="restrict",
+    )
+
+
+class PurchaseOrderCancelReason(models.Model):
```
Comment (sbejaoui) [inline purchase_cancel_reason/model/purchase_order.py]: please make dedicate file per model

File: purchase_cancel_reason/views/purchase_order.xml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                        name="cancel_reason_id"
+                        class="oe_inline"
+                        options="{'no_open': True}"
+                    />
+                </h5>
+            </xpath>
+        </field>
+    </record>
+
+    <record id="view_purchase_order_cancel_reason_form" model="ir.ui.view">
```
Comment (sbejaoui) [inline purchase_cancel_reason/views/purchase_order.xml]: same here

File: purchase_cancel_reason/wizard/purchase_cancel_reason.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# © 2013 Guewen Baconnier, Camptocamp SA
+# Copyright 2017 Okia SPRL
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo import _, fields, models
+from odoo.exceptions import UserError
+
+
+class PurchaseOrderCancel(models.TransientModel):
```
Comment (sbejaoui) [inline purchase_cancel_reason/wizard/purchase_cancel_reason.py]: the file name don't match the model name

File: purchase_cancel_reason/wizard/purchase_cancel_reason.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    )
+
+    def confirm_cancel(self):
+        self.ensure_one()
+        act_close = {"type": "ir.actions.act_window_close"}
+        purchase_ids = self._context.get("active_ids")
+        if purchase_ids is None:
+            return act_close
+        if len(purchase_ids) > 1:
+            raise UserError(_("Only 1 purchase ID expected"))
```
Comment (sbejaoui) [inline purchase_cancel_reason/wizard/purchase_cancel_reason.py]:
```suggestion
            raise UserError(_("Only 1 purchase expected"))
```

File: purchase_cancel_reason/wizard/purchase_cancel_reason.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+    reason_id = fields.Many2one(
+        comodel_name="purchase.order.cancel.reason", string="Reason", required=True
+    )
+
+    def confirm_cancel(self):
+        self.ensure_one()
+        act_close = {"type": "ir.actions.act_window_close"}
+        purchase_ids = self._context.get("active_ids")
+        if purchase_ids is None:
```
Comment (sbejaoui) [inline purchase_cancel_reason/wizard/purchase_cancel_reason.py]:
```suggestion
        if purchase_ids is None or self._context.get("active_model") != "purchase.order":
```

File: purchase_cancel_reason/wizard/purchase_cancel_reason_view.xml [hunk 8]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="view_purchase_order_cancel" model="ir.ui.view">
+        <field name="name">Reason for the cancellation</field>
+        <field name="model">purchase.order.cancel</field>
```
Comment (sbejaoui) [inline purchase_cancel_reason/wizard/purchase_cancel_reason_view.xml]: same here, the file name don't match model name

**Review by sbejaoui — CHANGES_REQUESTED** [review]
Small fixes needed,

File: purchase_cancel_reason/model/purchase_order.py (hunk 1 above)
Comment (jbaudoux) [reply purchase_cancel_reason/model/purchase_order.py]: Cancelation with one L then as we use US english

File: purchase_cancel_reason/readme/CONTRIBUTORS.rst [hunk 9]
```diff
@@ -0,0 +1,7 @@
+* Guewen Baconnier, Camptocamp SA
+* Sylvain Van Hoof <sylvain@okia.be>
+
+* `Ecosoft <http://ecosoft.co.th>`_:
+
+  * Kitti U. <kittiu@ecosoft.co.th>
+  * Tharathip C. <tharathipc@ecosoft.co.th>
```
Comment (jbaudoux) [inline purchase_cancel_reason/readme/CONTRIBUTORS.rst:7]: reformat list: remove blank lines and remove indent

File: purchase_cancel_reason/wizard/purchase_cancel_reason.py (hunk 6 above)
Comment (jbaudoux) [reply purchase_cancel_reason/wizard/purchase_cancel_reason.py]: I would even drop this restriction

**Comment by rousseldenis** [comment]
@pierre-halleux Could you attend comments ?

File: purchase_cancel_reason/readme/CONTRIBUTORS.rst (hunk 9 above)
Comment (rousseldenis) [reply purchase_cancel_reason/readme/CONTRIBUTORS.rst:7]: @jbaudoux This should remains as is as they are regrouped into the same company.

Looks like:

image

Comment (jbaudoux) [reply purchase_cancel_reason/readme/CONTRIBUTORS.rst:7]: ah ok :)

Comment (pierre-halleux) [reply purchase_cancel_reason/readme/CONTRIBUTORS.rst:7]: I reverted this commit so it looks like it used to

**Review by rousseldenis — APPROVED** [review]
Code review.

Nitpicking: Missing a space between field label and value:

image

**Comment by rousseldenis** [comment]
@sbejaoui @jbaudoux Could you update your review ?

**Review by jbaudoux — APPROVED** [review]

**Review by FrancoMaxime — COMMENTED** [review]
Same minor comment as @rousseldenis and @lmignon:
The purchase view can be improved

**Comment by lmignon** [comment]
@HviorForgeFlow CAn you merge this one since you merged the one for v17 based on this.

**Comment by lmignon** [comment]
> /ocabot merge nobump

Thank you!

**Pull request merged**